server/src/main/java/org/elasticsearch/action/bulk/BulkOperation.java

```diff
@@ -168,8 +180,63 @@ private void doRedirectFailures() {
         )) {
             return;
         }
-        Map<ShardId, List<BulkItemRequest>> requestsByShard = drainAndGroupRedirectsByShards(clusterState);
-        executeBulkRequestsByShard(requestsByShard, clusterState, this::completeBulkOperation);
+        Runnable runnable = () -> {
+            // Get new cluster state that includes any potential failure store rollovers.
+            var rolledOverState = observer.setAndGetObservedState();
+            Map<ShardId, List<BulkItemRequest>> requestsByShard = drainAndGroupRedirectsByShards(rolledOverState);
+            executeBulkRequestsByShard(requestsByShard, rolledOverState, this::completeBulkOperation);
+        };
+        rollOverFailureStores(runnable);
+    }
+
+    /**
+     * Send rollover requests for all failure stores that need it. After all requests have completed, we execute the given runnable.
+     * Any failures while rolling over will be added to the {@link BulkItemResponse} entries of the index requests that were redirected to
+     * the failure store that failed to roll over.
+     */
+    private void rollOverFailureStores(Runnable runnable) {
+        // Skip allocation of some objects if we don't need to roll over anything.
+        if (failureStoresToBeRolledOver.isEmpty()) {
+            runnable.run();
+            return;
+        }
+        try (RefCountingRunnable refs = new RefCountingRunnable(runnable)) {
+            for (String failureStore : failureStoresToBeRolledOver) {
```

Comment:
Yep, that's right. Are you suggesting just renaming the `failureStore` variable to `dataStream`, or also `failureStoresToBeRolledOver` to `dataStreamsToBeRolledOver`? I don't think I'm a fan of the latter. and the former might be a bit confusing because of the switch from `failureStoresToBeRolledOver` to `dataStream`. Maybe something like `failureStoreDataStreamName`? Although that's a bit too length IMO

Reply:
I was considering to use `dataStream` or `dataStreamName` just here. `failureStoresToBeRolledOver` can stay because as you said it will cause conflicts.